git_machete/cmd.py

```diff
@@ -339,409 +339,1171 @@ def popen_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs:
+                         "which are not local branches (perhaps they have been deleted?).\n" +
+                         "Slide them out from the definition file?" +
+                         pretty_choices("y", "e[dit]", "N"), opt_yes_msg=None)
+
+        def recursive_slide_out_invalid_branches(b: str) -> List[str]:
+            new_down_branches = flat_map(recursive_slide_out_invalid_branches, self.down_branches.get(b, []))
+            if b in invalid_branches:
+                if b in self.down_branches:
+                    del self.down_branches[b]
+                if b in self.annotations:
+                    del self.annotations[b]
+                if b in self.up_branch:
+                    for d in new_down_branches:
+                        self.up_branch[d] = self.up_branch[b]
+                    del self.up_branch[b]
+                else:
+                    for d in new_down_branches:
+                        del self.up_branch[d]
+                return new_down_branches
             else:
-                for d in new_down_branches:
-                    del up_branch[d]
-            return new_down_branches
+                self.down_branches[b] = new_down_branches
+                return [b]
+
+        self.roots = flat_map(recursive_slide_out_invalid_branches, self.roots)
+        self.managed_branches = excluding(self.managed_branches, invalid_branches)
+        if ans in ('y', 'yes'):
+            self.save_definition_file()
+        elif ans in ('e', 'edit'):
+            edit(cli_ctxt)
+            self.read_definition_file(cli_ctxt, verify_branches)
+
+    def render_tree(self) -> List[str]:
+        if not self.indent:
+            self.indent = "\t"
+
+        def render_dfs(b: str, depth: int) -> List[str]:
+            self.annotation = f" {self.annotations[b]}" if b in self.annotations else ""
+            res: List[str] = [depth * self.indent + b + self.annotation]
+            for d in self.down_branches.get(b, []):
+                res += render_dfs(d, depth + 1)
+            return res
+
+        total: List[str] = []
+        for r in self.roots:
+            total += render_dfs(r, depth=0)
+        return total
+
+    @staticmethod
+    def back_up_definition_file() -> None:
+        shutil.copyfile(definition_file_path, definition_file_path + "~")
+
+    def save_definition_file(self) -> None:
+        with open(definition_file_path, "w") as f:
+            f.write("\n".join(self.render_tree()) + "\n")
+
+    # Allowed parameter values for show/go command
+    @staticmethod
+    def allowed_directions(allow_current: bool) -> str:
+        current = "c[urrent]|" if allow_current else ""
+        return current + "d[own]|f[irst]|l[ast]|n[ext]|p[rev]|r[oot]|u[p]"
+
+    # Parse and evaluate direction against current branch for show/go commands
+    def parse_direction(self, cli_ctxt: CommandLineContext, param: str, b: str, allow_current: bool, down_pick_mode: bool) -> str:
```

Comment:
Hmmm I'd say a method like this one (also `allowed_directions`) probably doesn't belong well to `MacheteContext`... this is rather part of the implementation of command-line parsing.

Imagine for instance that we had a GUI client for git-machete written in Python (like, as if https://github.com/VirtusLab/git-machete-intellij-plugin was written in Python instead of Java/Kotlin). `MacheteContext` could be perfectly shared b/w CLI and GUI, but a method like `parse_direction` would be probably be useless for GUI.

Reply:
Generally, assume that `MacheteContext` contains things that can be used **by any client**, whether it is CLI, GUI, or even (not very realistically, but still) a mobile/web one.

Reply:
There are now some methods in gray area... like for instance the current implementation of `traverse`... can we consider it CLI-specific, or rather more generic? things like that will require further deliberation 🤔

Reply:
So what I'd recommend is to move the methods that are **clearly CLI-specific**, like `parse_direction`, out of `MacheteContext`. In the following PRs, we'll think how to redesign this behemoth of a code that is cmd.py so that it's elegantly split into layers (classes/maybe packages)